Adopting Existing Cloud Resources into Pulumi (#1556)
pulumi/docs: content/blog/adopting-existing-cloud-resources-into-pulumi/index.md

---
title: "Adopting Existing Cloud Resources into Pulumi"
date: "2019-08-15"

> meta_desc: "Adopt existing cloud infrastructure from Terraform, CloudFormation, ARM or YAML into Pulumi using `import`."

Reviewer: ```suggestion
meta_desc: "Migrating from Terraform or another IaC solution?  Learn how to adopt existing cloud infrastructure from Terraform, CloudFormation, ARM, YAML or other solutions into Pulumi using `import`."
```

meta_image: "console.png"
authors: ["luke-hoban"]
tags: ["Infrastructure"]
---

> Most cloud infrastrucutre projects involve working with existing cloud resources - either building on top of existing resources or adopting existing resources under management of a new and more robust infrastructure provisioning solution.

Reviewer: ```suggestion
Most cloud infrastrucutre projects involve working with existing cloud resources - either building on top of existing resources or adopting existing resources under management with a new and more robust infrastructure provisioning solution.
```

## Referencing Existing Resources
For referencing existing resoruces, Pulumi offers several tools.  

* The `.get` methods available on every resource let you [get all the details for a resource](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/aws/ec2/#Vpc-get) from the cloud provider based just on it's `id`.
* The `StackReference` resource lets you reference outputs of another stack for use as inputs to another stack, which is very useful for [organizing projects and stack](https://www.pulumi.com/docs/reference/organizing-stacks-projects/). 
* The [`terraform.state.RemoteStateReference()`](https://www.pulumi.com/blog/using-terraform-remote-state-with-pulumi/), [`aws.cloudformation.getStack()`](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/aws/cloudformation/#getStack) and [`azure.core.TemplateDeployment.get()`](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/azure/core/#TemplateDeployment-get) let you reference outputs from existing Terraform, CloudFormation and ARM deployments respectively.

Together, these make it easy to reference existing infrastructure regardless of how it was provisioned.

## Adopting Existing Resources

> For adopting resources, Pulumi offers the [`import`](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/azure/core/#TemplateDeployment-get) resource option to request that a resource defined in your Pulumi program adopt an existing resource in the cloud provider instead of creating a new one.  In keeping with it's focus on infrastructure as *code*, Pulumi lets you specify this `import` behaviour inside the Pulumi code for you infrastructure deployment, instead of outside it in some manual workflow.  In it's simplest form looks like this:

Reviewer: ```suggestion
For adopting existing resources, Pulumi offers the [`import`](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/azure/core/#TemplateDeployment-get) resource option to request that a resource defined in your Pulumi program adopt an existing resource in the cloud provider instead of creating a new one.  In keeping with it's focus on infrastructure as *code*, Pulumi lets you specify this `import` behavior inside the Pulumi code for you infrastructure deployment, instead of outside it in some manual workflow.  In it's simplest form looks like this:
```

```ts
const myVpc = new aws.ec2.Vpc("my-vpc", {
    cidrBlock: "10.11.0.0/16",
}, { import: "vpc-02cd0cd2d807f340c"});
```

When Pulumi first sees a resource (in this case `my-vpc`), if it has an `import` option set, it will adopt the existing resource specified by that `import` ID from the cloud provider under management instead of creating a new resource.

We can see this when we deploy our Pulumi program with `pulumi up`:

```
$ pulumi up
Previewing update (dev):

     Type                 Name               Plan       
     pulumi:pulumi:Stack  importtestaws-dev             
 =   └─ aws:ec2:Vpc       my-vpc             import     
 
Resources:
    = 1 to import
    1 unchanged
```

> Because the `import` is provided in code, it can be configured in many different ways.  We've seen users look up import ids from Pulumi config instead of hardcoding into their programs, construct them from predicatable names based on existing program parameters, or even conditionally add the `import: <id>` property depending on whether they want to deploy a stack in "adopt" mode or in "create" mode.

Reviewer: ```suggestion
Because the `import` is provided in code, it can be configured in several different ways.  You can look up import ids from Pulumi config instead of hardcoding them into programs, construct ids from predictable names based on program parameters, or even conditionally add the `import: <id>` property depending on whether one wants to deploy a stack in "adoption" or "creation" mode.
```

> This allows `import` to be used for a wide variety of adoption scenarios - from importing a single resource, to migrating a whole stack from an existing tool like Terraform to Pulumi, and even to automating the migration process across dozens of instances of an infrastructure deployment.

Reviewer: ```suggestion
`Import` can be used for a wide variety of adoption scenarios, from importing a single resource to migrating an entire stack from an existing tool like Terraform.  You can even automate an entire migration process across dozens of instances of infrastructure deployment.
```

### Walkthrough of Adopting Existing Infrastructure to Pulumi

To see one of these scenarios in action end-to-end, let's walk through the process of deploying some "existing" infrastructure outside of Pulumi, and then adopting it under management of a Pulumi program.  You can also check out the video below which walks through the same scenario.

<iframe width="560" height="315" src="https://www.youtube.com/embed/kX_3Wdft0Ms" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

We'll use Terraform to provision our "existing" infrastructure.  These resources could have been provisioned manually in the console, or using platform-native infrastructure-as-code solutions like CloudFormation, ARM or Kubernetes YAML.

> We'll also use TypeScript, though the process would work mostly the same in Python as well.

Reviewer: ```suggestion
We'll also use TypeScript, though of course, the process is similar regardless of the language you choose.
```

For our "existing" infrastructure, we'll use a small Azure deployment of a Virtual Machine and corresponding Resource Group and Networking resources.  We'll deploy [this example](https://www.terraform.io/docs/providers/azurerm/r/virtual_machine.html#example-usage-from-an-azure-platform-image-) using `terraform`.  Save the code in that example to `main.tf` and run `terraform init` and then `terraform apply` to deploy it.

#### 2. Create a Pulumi program

Next, we'll create a Pulumi program to describe the same infrastructure that exists in Azure.  We could write this by hand, but since we used Terraform to define the original infrastructure, we also have the option to use [`tf2pulumi`](https://github.com/pulumi/tf2pulumi) to automatically convert our `.tf` files to Pulumi.  

```bash
$ pulumi new azure-typescript --force
$ tf2pulumi < main.tf > index.ts
```

We now have a Pulumi program that describes the same infrastructure:

```ts
import * as pulumi from "@pulumi/pulumi";
import * as azure from "@pulumi/azure";

const config = new pulumi.Config();
const prefix = config.get("prefix") || "tfvmex";

const mainResourceGroup = new azure.core.ResourceGroup("main", {

>     location: "West US 2",

Reviewer: add a comment somewhere about region, or suggest config?

Author: This is the literal code emitted by tf2pulumi and a match for the TF example that was the seed for this. We can leave as is.

```

#### 4. Adopt resources

Now our preview shows that it will import the `ResourceGroup` instead:

```
$ pulumi up
Previewing update (dev2):

     Type                         Name              Plan       
 +   pulumi:pulumi:Stack          importazure-dev2  create     
 =   └─ azure:core:ResourceGroup  main              import     
 
Resources:
    + 1 to create
    = 1 to import
    2 changes
```

We can complete this process, adding `import: <id>` to all resources, and then import them all to Pulumi. 

```
Updating (dev2):

     Type                               Name              Status       
 +   pulumi:pulumi:Stack                importazure-dev2  created      
 =   ├─ azure:core:ResourceGroup        main              imported     
 =   ├─ azure:network:VirtualNetwork    main              imported     
 =   ├─ azure:network:Subnet            internal          imported     
 =   ├─ azure:network:NetworkInterface  main              imported     
 =   └─ azure:compute:VirtualMachine    main              imported     
 
Resources:
    + 1 created
    = 5 imported
    6 changes

Duration: 8s
```

#### 5. Managing the resources with Pulumi

> Once all the resources are adopted into Pulumi, we can now manage them by making incremental changes in Pulumi, and can discontinue using the existing `.tf` files to manage this set of infrastructure.  For example, if we want to add a new tag to our Virtual Machine, we can add it as below.  We can also now remove the `import` property if we want, or else leave it intact so long as the `id` of the resource remains the same (if in the future we need to make a change that replaces one of these imported resources, Pulumi will require that the `id` proeprty is removed, as it will no longer be correct). 

Reviewer: ```suggestion
Once all of the resources are adopted into Pulumi, we can manage them by making incremental changes in Pulumi, and can discontinue using the existing `.tf` files to manage this set of infrastructure.  For example, if we want to add a new tag to our Virtual Machine, we can use the code below.  We can also now remove the `import` property if we want, or else leave it intact so long as the `id` of the resource remains the same (if in the future we need to make a change that replaces one of these imported resources, Pulumi will require that the `id` property is removed, as it will no longer be correct.) 
```

```diff
const mainVirtualMachine = new azure.compute.VirtualMachine("main", {
    location: mainResourceGroup.location,
    name: `${prefix}-vm`,
    networkInterfaceIds: [mainNetworkInterface.id],
    osProfile: {

>         adminPassword: "Password1234!",

Reviewer: add comment about "only a sample" or make password "don't do this" ...

Author: Hmm - this again just comes literally from the Terraform example.  We could put a comment here I suppose (which wouldn't be real - as `tf2pulumi` wouldn't generate it.  Or else add an aside in the prose?

        adminUsername: "testadmin",
        computerName: "hostname",
    },
    osProfileLinuxConfig: {
        disablePasswordAuthentication: false,
    },
    resourceGroupName: mainResourceGroup.name,
    storageImageReference: {
        offer: "UbuntuServer",
        publisher: "Canonical",
        sku: "16.04-LTS",
        version: "latest",
    },
    storageOsDisk: {
        caching: "ReadWrite",
        createOption: "FromImage",
        managedDiskType: "Standard_LRS",
        name: "myosdisk1",
    },
    tags: {
        environment: "staging",
+       managedBy: "Pulumi",
    },
    vmSize: "Standard_DS1_v2",
});
```

Deploying this, Pulumi shows we will update the existing resource to add the new tag:

```
       ~ tags: {
          + managedBy: "Pulumi"
        }
```

> Since the resources are now under management of Pulumi, we can navigate over to the Pulumi Console to see our resources:

Reviewer: ```suggestion
Since the resources are now being managed by Pulumi, we can use the Pulumi Console to see our resources:
```

![The Pulumi Console stack resources page](./console.png)

> And from there, navigate directly to the Virtual Machine in the Azure Portal, including the newly added `managedBy: Pulumi` tag.

Reviewer: ```suggestion
And then, navigate to the Virtual Machine instance in the Azure Portal, and see the newly added `managedBy: Pulumi` tag.
```

![The Azure Portal virtual machine page](./portal.png)

> The result is that we have migrated our existing infrastructure into Pulumi and can now use all the rich features of Pulumi infrastructure-as-code to manage it going forward!

Reviewer: ```suggestion
The result is that we have migrated our existing infrastructure into Pulumi and can now use all the rich features of Pulumi to manage it going forward!
```
